src/certificate.rs

```diff
@@ -606,97 +240,222 @@ pub(crate) fn write_certificate(
     txn.save_object(object_id, &buf[..offset])
 }
 
-mod read_pki {
-    use der_parser::{
-        asn1_rs::Any,
-        ber::BerObjectContent,
-        der::{parse_der_integer, parse_der_sequence_defined_g, DerObject},
-        error::BerError,
+pub mod yubikey_signer {
+    //! Signer implementation for yubikey
+
+    use crate::{
+        error::{Error, Result},
+        piv::AlgorithmId,
+        piv::{sign_data, SlotId},
+        YubiKey,
+    };
+    use der::{
+        asn1::{Any, OctetString},
+        oid::db::rfc5912,
+        Encode, Sequence,
     };
-    use nom::{combinator, sequence::pair, IResult};
-    use rsa::{BigUint, RsaPublicKey};
-
-    use super::{OID_NIST_P256, OID_NIST_P384};
-    use crate::{piv::AlgorithmId, Error, Result};
-
-    /// From [RFC 8017](https://tools.ietf.org/html/rfc8017#appendix-A.1.1):
-    /// ```text
-    /// RSAPublicKey ::= SEQUENCE {
-    ///     modulus           INTEGER,  -- n
-    ///     publicExponent    INTEGER   -- e
-    /// }
-    /// ```
-    pub(super) fn rsa_pubkey(encoded: &[u8]) -> Result<RsaPublicKey> {
-        fn parse_rsa_pubkey(i: &[u8]) -> IResult<&[u8], (DerObject<'_>, DerObject<'_>), BerError> {
-            parse_der_sequence_defined_g(|i, _| pair(parse_der_integer, parse_der_integer)(i))(i)
+    use sha2::{Digest, Sha256, Sha384};
+    use signature::{Keypair, SignatureEncoding};
+    use std::{cell::RefCell, fmt, io::Write, marker::PhantomData};
+    use x509_cert::spki::{
+        self, AlgorithmIdentifierOwned, DynSignatureAlgorithmIdentifier, EncodePublicKey,
+        SubjectPublicKeyInfoRef,
+    };
+
+    type SigResult<T> = core::result::Result<T, signature::Error>;
+
+    /// Key to be used to sign certificates
+    pub trait KeyType {
+        /// Error returned when working with signature
+        type Error: Into<signature::Error> + fmt::Debug;
+        /// The signature type returned by the signer
+        type Signature: SignatureEncoding
+            + for<'s> TryFrom<&'s [u8], Error = Self::Error>
+            + fmt::Debug;
+        /// The public key used to verify signature
+        type VerifyingKey: EncodePublicKey
+            + DynSignatureAlgorithmIdentifier
+            + Clone
+            + From<Self::PublicKey>;
+        /// Public key type used to load the SPKI formatted key
+        type PublicKey: for<'a> TryFrom<SubjectPublicKeyInfoRef<'a>, Error = spki::Error>;
+
+        /// The algorithm used when talking with the yubikey
+        const ALGORITHM: AlgorithmId;
+
+        /// Prepare buffer before submitting it for signature
+        fn prepare(input: &[u8]) -> SigResult<Vec<u8>>;
```

Comment:
This is fine for now but it seems like it could potentially be replaced by [`PrehashSignature`](https://docs.rs/signature/latest/signature/trait.PrehashSignature.html)

Reply:
I'm not sure what you have in mind with `PrehashSignature`, but otherwise, this kind of relate to my ealier comment about getting rid of `AlgorithmId` altogether I think.

Reply:
I guess it won't cover all of the cases, never mind